CONTRIBUTING.txt

```diff
@@ -0,0 +1,98 @@
+    3. The user or contributor should write the issue by describing the problem they face or observe. Links to the forum or other resources are permitted but the issue should be complete and accurate and should not require the reader to visit the forum or any other platform to understand what is being described.
+    4. Issues should strive to describe the minimum acceptable condition.
+    5. Issues should focus on user tasks and avoid comparisons to other software solutions.
+    6. The user or Contributor SHOULD seek consensus on the accuracy of their observation and the value of solving the problem.
+    7. To submit a solution to a problem, a Contributor SHALL create a pull request back to the project.
+    8. Contributors and Maintainers SHALL NOT commit changes directly to the target branch.
+    9. To discuss a proposed solution, users MAY comment on the platform pull request. Forum conversations regarding the solution should be discouraged and conversation redirected to the Pull Request or the related issue.
+    10. To accept or reject a Pull Request, a Maintainer SHALL use the platform interface.
+    11. Maintainers SHOULD NOT merge their own PRs except:
+      - in exceptional cases, such as non-responsiveness from other Maintainers for an extended period (more than 1-2 days).
+      - If the maintainer is also the primary developer of the workbench or subsystem.
+
+    12. Maintainers SHALL merge valid PRs from other Contributors rapidly.
+    13. Maintainers MAY, at their discretion merge PRs that have not met all criteria to be considered valid to:
+        - (a) ending fruitless discussions
+        - (b) capturing toxic patches in the historical record
+        - (c) engaging with the Contributor on improving their patch quality.
+    14. Maintainers SHALL NOT make value judgments on correct patches.
+    15. Any Contributor who has value judgments on a PR SHOULD express these via their own PR.
+    16. The user who created an issue SHOULD close the issue after checking the PR is successful.
+    17. Maintainers SHOULD close user issues that are left open without action or update for an uncomfortable period.
+
+7. Branches and Releases
+
+    1. The project SHALL have one branch (“master”) that always holds the latest in-progress version and SHOULD always build.
+    2. The project SHALL NOT use topic branches for any reason. Personal forks MAY use topic branches.
+    3. To make a stable release a Maintainer shall tag the repository. Stable releases SHALL always be released from the repository master.
+
+
+8. Project Administration
+
+    1. The project administrators will manage the set of project Maintainers.  They SHALL maintain a sufficiently large pool of maintainers to ensure their succession and permit timely review of contributions.
+    2. A new Contributor who has had more than one PR accepted and who clearly understands the project goals and this process SHOULD be invited to become a Maintainer.
```

Comment:
This seems like a strange standard -- a single PR isn't much of a requirement for someone who will have commit access to the repository, but on the other hand, how does someone demonstrate that they "understand the project goals"?

Reply:
I agree.  It's too low.  This too is from C4 and I left it as is to start a discussion.  
Realistically, what should be the criteria for inviting new maintainers?

Reply:
I think that this another judgement call of either the Administrators or some number of Maintainers.  I would suggest that there should be more than one person involved in the decision.  Qualifications could include how many PR's of what quality along with any discussions in either code reviews or forum posts.  The people making the decision need to be comfortable with their decision, which implies that they think that they can predict how the now Maintainer will behave.

Reply:
> Qualifications could include how many PR's of what quality along with any discussions in either code reviews or forum posts. 

Remember, the goal is not to make judgments about the value of the contribution or even the quality of the code.  The maintainers are supposed to be enforcing the process.  I know this is leaning toward "optimistic merging" and we're agreed that we're not doing that.  OTOH, we shouldn't be setting the maintainer bar based on the ability to judge code.  That reinforces the idea that code quality is the most important idea.

I would rather see the maintainers focus on the process.  They would look at a PR and invite specific experts to do code-review and/or comment.  Then the maintainers would merge when they feel like all important concerns have been addressed.

Reply:
Somehow the people making the decision need to judge if the proposed Maintainer would do the job reasonably.  I agree that the Maintainer's job is not to make value judgements or judge the quality of the code unless they are one of the "specific experts" you mentioned.  Other than their interactions with other people in FreeCAD, what information is there?

Reply:
A referee does not have to be an expert player, or even a player at all. His job is purely to independently enforce and facilitate the rules of the game.

I was going to add "and resolve disputes", but I don't know if that is best?

Reply:
> I would rather see the maintainers focus on the process. They would look at a PR and invite specific experts to do code-review and/or comment.

Wouldn't this imply they are well connected within the FreeCAD community and know who happens to be an expert for a specific PR? So requirements are more on social aspects and networking with some basic understanding of coding.

Reply:
Yes, I suppose in fact it's not really necessary for them to be a "Contributor" at all -- we are looking for people who are solid community participants, who seem to understand our project goals, who know who to ask for reviews, and who understand this process we are developing. My read of C4 is that they were trying to maximize the number of _Maintainers_ but that the true goal is faster merging. What we are really going to need under the proposed structure is more _reviewers_. The work of a Maintainer is simple if you have a cadre of reviewers you trust.

Reply:
> Other than their interactions with other people in FreeCAD, what information is there?

Nothing that matters.  Ultimately, it's a judgment decision by the administrators.

Reply:
I would say trust and involvement with the community are important traits for a maintainer as well as sufficient social skills (aka not being a jerk) and sufficient coding skills. I have merged PRs and I think I understand the general objective of freecad (scriptable multiplatform FOSS parametric modelling) but I don't think I should be invited as maintainer at this stage.

Reply:
> I don't think I should be invited as maintainer at this stage.

I would have no qualms about inviting you because I honestly believe you have the best interests of the project in mind. 
I would also have no qualms about asking you to step back if it didn't work out. (I hope someone would do this for me in the same situation) 
Most importantly, I would respect someone for saying, "thanks but no thanks.  I'll contribute in other ways"

Reply:
It seems I did not read the full conversation before posting my comment, I was assuming maintainer==reviewer, if the task is mostly about enforcing the process then I agree the threshold of required technical knowledge can be **much** lower. Trust, engagement and commitment to the project are then more important.

Reply:
It seems I did not read the full conversation before posting my comment, I was assuming maintainer==reviewer, if the task is mostly about enforcing the process then I agree the threshold of required technical knowledge can be much lower. Trust, engagement and commitment to the project are then more important.

Reply:
Maybe something more realistic is that Administrator (with help of currently active Maintainers) shall periodically (every 3 months ?) review last 12-month contributions and decide if new maintainers can be invited (I'd like to have criteria but I have no idea).

Also worth mentioning that maintainers are permanently encouraged to act in the reasonable limit of their skills and competences, and to call for other reviewers in case it seems needed ?

Reply:
Are you proposing a text change?

Reply:
> Are you proposing a text change?

ATM just opening a discussion based on a proposal to slightly change the process. I we can agree on that, we shall change the text accordingly. :)
If it may ease discussion, I can write it under a text change suggestion.

Reply:
Interesting discussion. All the discussion re the qualities of the maintainer may be over-worrying the problem. A maintainer is by definition not an interested party ie must not have stake in the outcome. The rest is largely mechanical. This is a very old principle in law, sport, etc. The trust relationship then becomes more like employer - employee.

Reply:
I think what's needed to be a good maintainer is part technical (understand how the freecad code works and how this process works), and part human (understand how to cope with the human side of the process and basically help other developers). But that's hard to put into text...

Reply:
The text has been updated.  Resolving since there's no additional discussion.  Re-open if I missed anything.